File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectComboFragment.java
package com.cinema.ticket_booking.ui.booking;

import android.os.Bundle;
import android.util.Log;
import android.view.*;
import android.widget.Toast;
import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.ComboResponse;
import com.cinema.ticket_booking.databinding.FragmentSelectComboBinding;
import com.cinema.ticket_booking.data.model.response.BookingResponse;
import com.cinema.ticket_booking.util.Resource;
import com.cinema.ticket_booking.util.SnackbarHelper;

import dagger.hilt.android.AndroidEntryPoint;
import java.util.List;

@AndroidEntryPoint
public class SelectComboFragment extends Fragment {

    private FragmentSelectComboBinding binding;
    private SelectComboViewModel viewModel;
    private ComboAdapter adapter;

    @Override
    public View onCreateView(@NonNull LayoutInflater i, ViewGroup c, Bundle s) {
        binding = FragmentSelectComboBinding.inflate(i, c, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(SelectComboViewModel.class);

        binding.btnBack.setOnClickListener(v -> Navigation.findNavController(view).popBackStack());
        binding.tvMovieTitle.setText(SelectShowtimeViewModel.pendingMovieTitle);

        setupRecyclerView();
        observeViewModel();
        setupClickListeners();
    }

    private void setupRecyclerView() {
        binding.rvCombos.setLayoutManager(new LinearLayoutManager(requireContext()));
    }

    private void observeViewModel() {
        viewModel.getCombos().observe(getViewLifecycleOwner(), resource -> {
            binding.progressBar.setVisibility(resource.isLoading() ? View.VISIBLE : View.GONE);
            if (resource.isSuccess() && resource.data != null) {
                adapter = new ComboAdapter(resource.data,
                        comboId -> {
                            viewModel.addCombo(comboId);
                            updateItemSelectedCount();
                        },
                        comboId -> {
                            viewModel.removeCombo(comboId);
                            updateItemSelectedCount();
                        },
                        viewModel.getSelectedCombos());
                binding.rvCombos.setAdapter(adapter);
                updateItemSelectedCount();
            } else if (resource.isError()) {
                SnackbarHelper.showError(binding.getRoot(), resource.message);
            }
        });

        // Instant Total Calculation (Optimistic UI)
        viewModel.getLocalTotal().observe(getViewLifecycleOwner(), total -> {
            binding.tvTotal.setText(String.format(java.util.Locale.getDefault(), "%,.0fđ", total));
        });
    }

    private void setupClickListeners() {
        binding.btnSkip.setOnClickListener(v -> handleContinue());
        binding.btnContinue.setOnClickListener(v -> handleContinue());
    }

    private void handleContinue() {
        // Show loading state for the final transition quote
        showLoading(true);

        viewModel.getFinalQuote().observe(getViewLifecycleOwner(), resource -> {
            if (resource.status == Resource.Status.SUCCESS && resource.data != null) {
                navigateToConfirm(resource.data);
            } else if (resource.status == Resource.Status.ERROR) {
                showLoading(false);
                SnackbarHelper.showError(binding.getRoot(),
                        resource.message != null ? resource.message : "Lỗi tính toán giá cuối cùng");
            }
        });
    }

    private void navigateToConfirm(BookingResponse quote) {
        Bundle bundle = new Bundle();
        if (getArguments() != null) {
            bundle.putLong("expireTime", getArguments().getLong("expireTime"));
        }

        // Pass the Parcelable quote to the next screen
        bundle.putParcelable("initialQuote", quote);

        NavHostFragment.findNavController(this)
                .navigate(R.id.action_selectCombo_to_confirmBooking, bundle);
    }

    private void showLoading(boolean isLoading) {
        binding.progressBar.setVisibility(isLoading ? View.VISIBLE : View.GONE);
        binding.btnContinue.setEnabled(!isLoading);
        binding.btnSkip.setEnabled(!isLoading);
        binding.btnBack.setEnabled(!isLoading);

        if (isLoading) {
            binding.btnContinue.setAlpha(0.6f);
            binding.btnContinue.setText("Đang xử lý...");
        } else {
            binding.btnContinue.setAlpha(1.0f);
            binding.btnContinue.setText("Tiếp tục");
        }
    }

    private void updateItemSelectedCount() {
        int items = 0;
        for (Integer qty : viewModel.getSelectedCombos().values()) {
            items += qty;
        }
        binding.tvItemSelected.setText("(" + items + " phần)");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectComboFragment.kt
package com.cinema.ticket_booking.ui.booking

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.BookingResponse
import com.cinema.ticket_booking.databinding.FragmentSelectComboBinding
import com.cinema.ticket_booking.util.Resource
import com.cinema.ticket_booking.util.SnackbarHelper
import dagger.hilt.android.AndroidEntryPoint
import java.util.Locale

@AndroidEntryPoint
class SelectComboFragment : Fragment() {

    private var _binding: FragmentSelectComboBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: SelectComboViewModel
    private var adapter: ComboAdapter? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSelectComboBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[SelectComboViewModel::class.java]

        binding.btnBack.setOnClickListener {
            Navigation.findNavController(view).popBackStack()
        }
        binding.tvMovieTitle.text = SelectShowtimeViewModel.pendingMovieTitle

        setupRecyclerView()
        observeViewModel()
        setupClickListeners()
    }

    private fun setupRecyclerView() {
        binding.rvCombos.layoutManager = LinearLayoutManager(requireContext())
    }

    private fun observeViewModel() {
        viewModel.combos.observe(viewLifecycleOwner) { resource ->
            binding.progressBar.visibility = if (resource.isLoading) View.VISIBLE else View.GONE
            if (resource.isSuccess && resource.data != null) {
                adapter = ComboAdapter(
                    resource.data,
                    { comboId ->
                        viewModel.addCombo(comboId)
                        updateItemSelectedCount()
                    },
                    { comboId ->
                        viewModel.removeCombo(comboId)
                        updateItemSelectedCount()
                    },
                    viewModel.selectedCombos
                )
                binding.rvCombos.adapter = adapter
                updateItemSelectedCount()
            } else if (resource.isError) {
                SnackbarHelper.showError(binding.root, resource.message ?: "Lỗi tải combo")
            }
        }

        // Instant Total Calculation (Optimistic UI)
        viewModel.localTotal.observe(viewLifecycleOwner) { total ->
            binding.tvTotal.text = String.format(Locale.getDefault(), "%,.0fđ", total)
        }
    }

    private fun setupClickListeners() {
        binding.btnSkip.setOnClickListener { handleContinue() }
        binding.btnContinue.setOnClickListener { handleContinue() }
    }

    private fun handleContinue() {
        // Show loading state for the final transition quote
        showLoading(true)

        viewModel.getFinalQuote().observe(viewLifecycleOwner) { resource ->
            if (resource.status == Resource.Status.SUCCESS && resource.data != null) {
                navigateToConfirm(resource.data)
            } else if (resource.status == Resource.Status.ERROR) {
                showLoading(false)
                SnackbarHelper.showError(
                    binding.root,
                    resource.message ?: "Lỗi tính toán giá cuối cùng"
                )
            }
        }
    }

    private fun navigateToConfirm(quote: BookingResponse) {
        val bundle = Bundle().apply {
            arguments?.let {
                putLong("expireTime", it.getLong("expireTime"))
            }
            putParcelable("initialQuote", quote)
        }

        NavHostFragment.findNavController(this)
            .navigate(R.id.action_selectCombo_to_confirmBooking, bundle)
    }

    private fun showLoading(isLoading: Boolean) {
        binding.progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
        binding.btnContinue.isEnabled = !isLoading
        binding.btnSkip.isEnabled = !isLoading
        binding.btnBack.setEnabled(!isLoading)

        if (isLoading) {
            binding.btnContinue.alpha = 0.6f
            binding.btnContinue.text = "Đang xử lý..."
        } else {
            binding.btnContinue.alpha = 1.0f
            binding.btnContinue.text = "Tiếp tục"
        }
    }

    private fun updateItemSelectedCount() {
        var items = 0
        for (qty in viewModel.selectedCombos.values) {
            items += qty
        }
        binding.tvItemSelected.text = "($items phần)"
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
